pcdsdevices/signal.py

```diff
@@ -100,3 +102,70 @@ def _run_sub_value(self, *args, **kwargs):
             if value != old_value or not self._update_only_on_change:
                 self._run_subs(sub_type=self.SUB_VALUE, obj=self, value=value,
                                old_value=old_value)
+
+
+class AvgSignal(Signal):
+    """
+    Signal that acts as a rolling average of another signal.
+
+    Parameters
+    ----------
+    signal: ``Signal``
+        Any subclass of ``ophyd.signal.Signal`` that returns a numeric value.
+        This signal will be subscribed to be `AvgSignal` to calculate the mean.
+
+    averages: ``int``
+        The number of SUB_VALUE updates to include in the average. New values
+        after this number is reached will begin overriding old values.
+    """
+    def __init__(self, signal, averages, *, name, parent=None, **kwargs):
+        super().__init__(name=name, parent=parent, **kwargs)
+        if isinstance(signal, str):
+            signal = getattr(parent, signal)
+        self.raw_sig = signal
+        self._lock = RLock()
+        self.averages = averages
+        self._con = False
+        t = Thread(target=self._init_subs, args=())
+        t.start()
+
+    def _init_subs(self):
+        self.raw_sig.wait_for_connection()
+        self.raw_sig.subscribe(self._update_avg)
+        self._con = True
+
+    @property
+    def connected(self):
+        return self._con
+
+    @property
+    def averages(self):
+        """
+        The size of the internal buffer of values to average over.
+        """
+        return self._avg
+
+    @averages.setter
+    def averages(self, avg):
+        """
+        Reinitialize an empty internal buffer of size ``avg``.
+        """
+        with self._lock:
+            self._avg = avg
+            self.index = 0
+            # Allocate uninitalized array
+            self.values = np.empty(avg)
+            # Fill with nan
+            self.values.fill(np.nan)
+
+    def _update_avg(self, *args, value, **kwargs):
+        """
+        Add new value to the buffer, overriding old values if needed.
+        """
+        with self._lock:
+            self.values[self.index] = value
+            self.index += 1
+            if self.index == len(self.values):
+                self.index = 0
```

Comment:
Instead of keeping track of a `self.index` you could always use `np.roll` and replace the first value.

```python
x = array([1, 2, 3, 4])
x = np.roll(x, shift=1)
x
array([4, 1, 2, 3])
```

Reply:
I didn't know about np.roll.
```
x = np.array([np.nan, np.nan, np.nan, np.nan])
for i in range(5):
    x = np.roll(x, shift=1)
    x[0] = i
    print(x)
[ 0. nan nan nan]
[ 1.  0. nan nan]
[ 2.  1.  0. nan]
[3. 2. 1. 0.]
[4. 3. 2. 1.]
```
Seems ok, but I read the source code and this allocates a new array every time. I'd much rather keep modifying the array in place because this callback might be run at 120Hz.

Reply:
Good point. The idea was that it would be nice if you could always tell the order of events but this is probably unnecessary especially at the cost of performance. You checked this against a 120Hz PV at all?

Reply:
I checked it in the `BeamStats` object and it seems ok, though python prints at 120Hz are surprisingly expensive. The stuttering is clearly from the gateway issue.